cli: fix pretty print alignment

pkit opened the pull request:
This patch modifies the `pretty` table format implementation so that
caller scan specify alignment of results inside the table. The
alignment directive supports "l", "r" and "c" for left, right and
center horizontal alignment, respectively.

Alignment in results printed by the `sql` client have alignment
decided automatically based on column type:

- strings and unicode: `left`
- ints and other numericals: `right`
- bools: `center`

This patch also adds proper alignment to `node ls` and other
"ad-hoc/non-table" queries.

Fixes #10042.

Release note (cli change): command-line utilities that print results
with the `pretty` format now use consistent horizontal alignment for
every row of the result.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18491)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=18491) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=18491) before we can accept your contribution.<br/><hr/>**Constantine Peresypkin** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=18491) it.</sub>

pkit commented:
Depends on upstream: https://github.com/olekukonko/tablewriter/pull/74

knz commented:
Many thanks for working on this. This problem had been bugging us for *ages*! cc @andreimatei who was regularly pushing me to look into this :smile_cat: 

I'll help with the upstream dependency a bit later. In the meantime, could you enhance this patch by supporting a string-based alignment specifier, instead of arrays of int constants from the tablewriter package? The idea would be as follows:

1) we want to restrict the dependency on the specific rendering package (`tablewriter`) to format_table.go. The rest of the source code should be agnostic wrt what rendering is being used.

2) the current code is excessively verbose.

The method I would like to use is the same as used in [LaTeX table alignment specifiers](https://en.wikibooks.org/wiki/LaTeX/Tables#The_tabular_environment), where each character in a string would specify the alignment for the corresponding result column. We'd support just "l", "c", "r", and "a" for tablewriter's "default" which really means "autodetected". 

Thank you again!

---

Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18491#-:-Ku48FN5fVYw5OY9Udx-:bg6t81v)*
<!-- Sent from Reviewable.io -->

knz commented:
Also please prefix your commit title with "`cli: `" Thanks!

knz commented:
Also please rebase your fix on top of #17816 - this significantly refactors the code and is likely to merge before this PR.

pkit commented:
Totally agree that dependency on tablewriter should be contained. Was not sure what's a better way to rewrite these alignment identifiers though. Using just one string seems good enough. Will use it.

---

Review status: 5 of 6 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18491#-:-Ku6N50JX5wA4X1ob6ew:bh0lshv)*
<!-- Sent from Reviewable.io -->

pkit commented:
Will rebase against #17816 when it lends in master.

knz commented:
It's ready.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

pkit commented:
Rebased. Code fragility was strong in this one. Needed to refactor the strange `io.EOF` handling

knz commented:
Okay this is better. The first commit is fine.
Now I think  the plumbing you have achieved in `format_table.go` has too many conditionals. In general conditionals in a "hot path" (like  the `Next` methods) is undesirable, so I propose the following:

- change the return type of `Align` in the interface `rowStrIter` to `string`
- make `newRowSliceIter` take the alignment string as argument and store it in `rowSliceIter` without processing, and make `(*rowSliceIter).Align()` return it.
- add a new field `lastRow []driver.Value` in `rowIter`
- leave the methods `getAllRowStrings` and `getNextRowStrings` alone as they were initially, instead, for `(*rowIter).Next()` *inline* the code from `getNextRowString` into `(*rowIter).Next()` and make it store the current row being retrieved into the rowIter's `lastRow`
- make `(*rowIter).Align()` compute the alignment string based on the last row stored in `rowIter`
- extend the interface `rowReporter.beforeFirstRow` to take an additional `alignment string` parameter. 
- change the `render` function accordingly. It is possible to call `(rowStrIter).Align()` at the point `beforeFirstRow` is called, because by then there will have been at least one row read by the iterator.
- Ignore the new `beforeFirstRow` argument in all implementations of the interface except for `prettyReporter`, where you use it to call `p.table.SetTableAlignment`. No need to store a new field in `prettyReporter`. 

---

Reviewed 6 of 6 files at r2, 6 of 6 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18491#-:-KuKBvsd6V3KNTuhokNt:b48u5cy)*
<!-- Sent from Reviewable.io -->

pkit commented:
What about `nil` handling? `driver.Value` is just an `interface{}`, can I safely derive real type from `nil`s?
What if we really get an `interface{}(nil)` from driver up until "normal" value is materialized there?

knz commented:
Oh, how unfortunate.
It's rather important to not be computing something in the `Next` method especially since most formatters will not use this information.

So what can we do here? It's frustrating to look at this knowing that *the pq driver does know about the column types from the beginning*! 

So I suggest the following. It's a bit more intricate but I am confident it is worth it:
- patch the copy of `pq` in the `vendor` sub-directory, to add a new method `func (rs*rows).GetColTypes() []oid.Oid { return rs.colTyps }`
- extend the `sqlRowsI` interface in `sql_util.go` with `GetColTypes() []oid.Oid`
- (check that this runs fine: the rows object returned by Query can be casted to `sqlRowsI` without a panic, that is, we are indeed seeing pq's rows object)
- use this new method in the `rowsIter.Align` method.

You may need to create a PR for the change in the `vendor` directory as this is handled by a separate repository. See `build/README.md` for details.

pkit commented:
Yeah, now we're talking. Checked out and indeed got almost +50% of time spent in `Next()` which is obviously a bad idea. Was frustrated to see that no type info is easily accessible anywhere. Time to fix it!

pkit commented:
Added https://github.com/cockroachdb/vendored/pull/8

pkit committed:
cli: fix pretty print alignment

This patch modifies the `pretty` table format implementation so that
caller scan specify alignment of results inside the table. The
alignment directive supports "l", "r" and "c" for left, right and
center horizontal alignment, respectively.

Alignment in results printed by the `sql` client have alignment
decided automatically based on column type:

- strings and unicode: `left`
- ints and other numericals: `right`
- bools: `center`

This patch also adds proper alignment to `node ls` and other
"ad-hoc/non-table" queries.

Fixes #10042.

Release note (cli change): command-line utilities that print results
with the `pretty` format now use consistent horizontal alignment for
every row of the result.

mjibson commented:
I've merged https://github.com/lib/pq/pull/647. You'll need to submit a PR to vendored with that change and verify it works for this.

pkit commented:
@mjibson @knz tests pass with current HEAD of https://github.com/lib/pq

knz commented:
All right. I think you can run `dep ensure --upgrade github.com/lib/pq` in the `vendor` sub-dir to just upgrade that dependency. Then submit a PR to our vendored repo.

---

Review status: 4 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18491#-:-KuLsc19aBcNLQZdzrhk:bdji8z1)*
<!-- Sent from Reviewable.io -->

pkit commented:
Did that already. The result is in https://github.com/cockroachdb/vendored/pull/9

---

Review status: 4 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18491#-:-KuNwadXjwq4SqQW-UQr:bku78mr)*
<!-- Sent from Reviewable.io -->

knz commented:
Ok you can bump the vendor hash now.

pkit commented:
I think https://github.com/olekukonko/tablewriter/pull/74 should land first. Then I can update all the hashes. And resubmit these commits.

knz commented:
Ack. Thanks

knz commented:
I think you have a green light now.

RaduBerinde commented:
Hi @pkit,

The vendor commit has been updated; you will have to rerun the update against the newer one.  You can then push the new commit to the `vendored` repo in a new feature branch (not `master`!). That branch will ensure the commit is accessible. Then, after this PR merges, we can `update vendored/master` to that new commit and delete the feature branch.

See this document for more details: https://github.com/cockroachdb/cockroach/tree/master/build#master-branch-pointer-in-vendored-repo

knz commented:
Technically, since you don't have permission to push to vendor directly, please create a new branch in your clone of the repo; I'll then copy your branch to the vendor repo so it becomes visible for CI. We'll merge the branch after this PR merges.

knz commented:
We have found a couple of bugs in our embedded copy of `dep`, which may (or may have not) caused issues when you tried to update the dependencies. These issues are fixed now, so you can probably safely retry now.

nvanbenschoten commented:
Ping! It would be nice to resolve the associated issue.

knz committed:
sql: fix a flaky test

Release note: none

knz commented:
Hi everyone! I have upgraded everything that needed to be upgraded and updated this PR.

@pkit I took the liberty to update your branch with a rebase on top of a more recent version of CockroachDB and fixed the merge conflict. The commit is still all yours!

Merging this when CI lets me.

knz commented:
@pkit many thanks for your contribution. This definitely increases UX.

jseldess commented:
@knz, is there any user-facing change here? I can't tell if there are new options to go along with `pretty`.

knz commented:
@jseldess I reckon that example outputs where you have mixed text/numeric values in tables could benefit from a re-generation to show the new alignment. There is no functional change other than the way the tables appear on the screen.

jseldess commented:
Got it. Thanks, @knz.
